### DISCUSSION TITLE
ActivityExtensions.SetStatus vs Activity.Status

### DISCUSSION THREAD

**Discussion by jamescrosswell** [open-telemetry/opentelemetry-dotnet#4703] [Q&A]
The docs for [ActivityExtensions.SetStatus](https://github.com/open-telemetry/opentelemetry-dotnet/blob/dacc532d51ca0f3775160b84fa6d7d9403a8ccde/src/OpenTelemetry.Api/Trace/ActivityExtensions.cs#L33C1-L48C6) indicate:
>Activity class in .NET does not support 'Status'.

And the implementation for that extension method just sets a couple of otel tags on the activity:
```csharp
        activity.SetTag(SpanAttributeConstants.StatusCodeKey, StatusHelper.GetTagValueForStatusCode(status.StatusCode));
        activity.SetTag(SpanAttributeConstants.StatusDescriptionKey, status.Description);
```

There is an [`Activity.Status` property in .NET 7](https://learn.microsoft.com/en-gb/dotnet/api/system.diagnostics.activity.setstatus?view=net-7.0). 

This is a bit confusing. The `Activity.Status` of the span/activity that you get from the following code is `ActivityStatusCode.Unset`... not what you'd expect:
```csharp
        using var activity = MyActivitySource.StartActivity("Funny business");
        try
        {
            throw new Exception("test activity.RecordException");
        }
        catch (Exception e)
        {
            activity?.RecordException(e);
            activity?.SetStatus(Status.Error.WithDescription(e.Message));
        }
```

Is this because the `Activity` class on older versions of the .NET Framework didn't have a `Status` field? 

I wonder if it makes sense to augment the implementation of `ActivityExtensions.SetStatus` with something like:
```
        public static void SetStatus(this Activity activity, Status status)
        {
            Debug.Assert(activity != null, "Activity should not be null");

#if NET7_0_OR_GREATER
            // Set Activity.Status directly on target frameworks that support this
            activity.SetStatus(status.ToActivityStatusCode(), status.message); 
#endif
            activity.SetTag(SpanAttributeConstants.StatusCodeKey, StatusHelper.GetTagValueForStatusCode(status.StatusCode));
            activity.SetTag(SpanAttributeConstants.StatusDescriptionKey, status.Description);
        }
```

`ToActivityStatusCode` could probably be avoided by writing an implicit conversion operator for `OpenTelemetry.Trace.Status`.

**Comment by vishweshbankwar** [comment]
@jamescrosswell If you are using the latest stable version of SDK, you can simply call `activity.SetStatus(ActivityStatusCode)` to set the status on activity (No need for conditional compilation).  Note that this API is available in library `System.Diagnostics.DiagnosticSource` starting version `6.0` [reference](https://github.com/open-telemetry/opentelemetry-dotnet/tree/dacc532d51ca0f3775160b84fa6d7d9403a8ccde/src/OpenTelemetry.Api#setting-status), it does not directly depend on the targetframework of your app.  Using `ActivityExtensions.SetStatus` is not recommended, we do have a gap on documentation and will get this fixed. Thanks for the feedback.

**Comment by jamescrosswell** [comment]
Thanks @vishweshbankwar... that makes sense. Maybe worth adding an `[Obsolete]` attribute to the `ActivityExtensions.SetStatus` extension method in OpenTelemetry?
